---
title: "Fine-mapping of Breast Cancer GWAS"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source('../trackplot.R')
source('../GWAS_funs_libs.R')
library(ggplot2)
library(kableExtra)
```

***

## Intro

Fine-mapping was done on BRCA GWAS summary statistics from the Breast Cancer Consortium (BCAC). Prior weights were obtained from TORUS. I selected all LD blocks that have an FDR $\leq 10$% based on the q-values from TORUS. I used `L = 1` for fine-mapping with `susie_rss()` which works on out-of-sample LD.

The SNP covariance matrix was estimated as follows as adviced by Yuxin:

$$ R = (X^TX + \hat{z}\hat{z}^T)/n$$

where $X$ is the $n \times p$ 1000 Genomes genotype matrix, which has been column centered and scaled to unit variance. The outer-product of z-scores $\hat{z}\hat{z}^T$ penalize the covariance when the signs of $\hat{z}$ dont agree. 

To get the LD correlation matrix $r_{ij}$, each element of $R$ is transformed as follows:

$$ r_{ij} = \frac{R_{ij}}{\sqrt{R_{ii}}\sqrt{R_{jj}}} $$

## Enrichment

```{r}
torus <- read.delim('../BRCA/torus_enrich.txt', header=F, sep="")
torus <- torus[-1,] # dont need interncept
torus$V1 <- factor(torus$V1, levels = torus$V1[order(torus$V2, decreasing = F)])

ggplot(torus, aes(x=V1, y=V2)) + 
  geom_point() + 
  coord_flip() +
  geom_hline(yintercept = 0, color="red", linetype="dashed") +
  geom_errorbar(aes(ymin=V3, ymax=V4), width=.2) +
  theme(text = element_text(size=14),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
  ylab('log Odds Ratio') + 
  xlab('Category')
```


```{r}

load('../BRCA/BRCA_susie_L1.Robj')
susie.res <- susie_res
load('../BRCA/BRCA_susie_L1_UNIFORM.Robj')
susie.res.unif <- susie_res

sumstats <- as_tibble(data.table::fread('~/GERMLINE/SUMMARY_STATISTICS/susie/BRCA_Sumstats_SuSiE_Ready.txt.gz', sep = '\t', header = T))

merged.sumstats <- merge_susie_sumstats(susie.res, sumstats)
merged.sumstats.unif <- merge_susie_sumstats(susie.res.unif, sumstats)
```

## Fine-mapping summary

### Variants per credible-set

```{r}
variants.per.chunk <- merged.sumstats %>% group_by(locus) %>% summarise(n = sum(CS))

hist(variants.per.chunk$n, xlab = 'Number of putative variants', main = 'Number of variants per CS', breaks = 50)
text(20, 10, paste('Median: ', round(median(variants.per.chunk$n),2)))
```

### SuSiE PIP before and after torus priors

```{r}
pip_df <- data.frame(pip = merged.sumstats$susie_pip, pip_uniform = merged.sumstats.unif$susie_pip)

ggplot(pip_df, aes(x=pip_uniform, y=pip)) + geom_point() + geom_abline(slope = 1, color='red') + xlab('PIP (uniform)') + ylab('PIP')
```

### Link to table

[Google sheets link for Breast Cancer finemapping](https://docs.google.com/spreadsheets/d/1tTEj1ZqtCV5CgPqVNz_FEpifNtTmu6vHuYA1aLZsRd8/edit?usp=sharing)

