---
title: "Fine-mapping of cancer GWAS"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source('../trackplot.R')
source('../GWAS_funs_libs.R')
library(ggplot2)
library(kableExtra)
```

***

Fine-mapping was done on several cancer types. Prior weights were obtained from TORUS. I selected all LD blocks that have an FDR $\leq 10$% based on the q-values from TORUS. I used `L = 1` for fine-mapping with `susie_rss()` which works on out-of-sample LD.

The SNP covariance matrix was estimated as follows as adviced by Yuxin:

$$ R = (X^TX + \hat{z}\hat{z}^T)/n$$

where $X$ is the $n \times p$ 1000 Genomes genotype matrix, which has been column centered and scaled to unit variance. The outer-product of z-scores $\hat{z}\hat{z}^T$ penalize the covariance when the signs of $\hat{z}$ dont agree. 

To get the LD correlation matrix $r_{ij}$, each element of $R$ is transformed as follows:

$$ r_{ij} = \frac{R_{ij}}{\sqrt{R_{ii}}\sqrt{R_{jj}}} $$

## Fine-mapping of breast cancer

### Torus enrichment estimates

```{r}
torus <- read.delim('../BRCA/torus_enrich.txt', header=F, sep="")
torus <- torus[-1,] # dont need interncept
torus$V1 <- factor(torus$V1, levels = torus$V1[order(torus$V2, decreasing = F)])

ggplot(torus, aes(x=V1, y=V2)) + 
  geom_point() + 
  coord_flip() +
  geom_hline(yintercept = 0, color="red", linetype="dashed") +
  geom_errorbar(aes(ymin=V3, ymax=V4), width=.2) +
  theme(text = element_text(size=14),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
  ylab('log Odds Ratio') + 
  xlab('Category')
```


```{r}

load('../BRCA/BRCA_susie_L1.Robj')
susie.res <- susie_res
load('../BRCA/BRCA_susie_L1_UNIFORM.Robj')
susie.res.unif <- susie_res

sumstats <- as_tibble(data.table::fread('~/GERMLINE/SUMMARY_STATISTICS/susie/BRCA_Sumstats_SuSiE_Ready.txt.gz', sep = '\t', header = T))

merged.sumstats <- merge_susie_sumstats(susie.res, sumstats)
merged.sumstats.unif <- merge_susie_sumstats(susie.res.unif, sumstats)
```

### Number of variants per block

```{r}
variants.per.chunk <- merged.sumstats %>% group_by(locus) %>% summarise(n = sum(CS))

hist(variants.per.chunk$n, xlab = 'Number of putative variants', main = 'Number of variants per CS', breaks = 50)
text(20, 10, paste('Median: ', round(median(variants.per.chunk$n),2)))
```

### SuSiE PIP before and after torus priors

```{r}
pip_df <- data.frame(pip = merged.sumstats$susie_pip, pip_uniform = merged.sumstats.unif$susie_pip)

ggplot(pip_df, aes(x=pip_uniform, y=pip)) + geom_point() + geom_abline(slope = 1, color='red') + xlab('PIP (uniform)') + ylab('PIP')
```

### Annotate SNPs with PIP > 0.1

Add functional annotation used in Torus

```{r}
top.snps <- merged.sumstats[merged.sumstats$susie_pip > 0.1 & merged.sumstats$CS == 1, ]

annot.df <- suppressMessages(vroom::vroom('~/GERMLINE/SUMMARY_STATISTICS/torus/BRCA_Torus_annotations.txt.gz', col_names = T))
annot.df <- annot.df[annot.df$snp %in% top.snps$snp, ]
colnames(annot.df) <- gsub(colnames(annot.df), pattern = '.bed_d', replacement = '') # clean up annotation names

annot.df$merged <- apply(annot.df, MARGIN = 1, FUN = function(x){return(paste0(names(x)[which(x==1)], collapse = ';\n'))})
top.snps <- inner_join(top.snps, annot.df[ ,c("snp","merged")], by = "snp")
```

Add gene innformation (nearby gene, drivers, immune genes)

```{r}
# add regular genes
genes <- suppressMessages(vroom::vroom('~/GERMLINE/refGenome/GENES_COORDs_b37.txt', col_names = T))
top.snps <- add.nearby.gene(top.snps, genes, dist=10000, gene_type='nearby_gene')

# add drivers
drivers <- readLines('~/GERMLINE/ANNOTATIONS/cancer_driver_genes/AllDrivers_FDR_0.10.txt')
top.snps <- add.nearby.gene(top.snps, genes[genes$gene %in% drivers, ], dist = 100000, gene_type='nearby_driver')
```

Add eQTL information

```{r}
gtex <- suppressMessages(vroom::vroom('~/GERMLINE/GTEx_Analysis_v7_eQTL/Breast_Mammary_var_genes.txt.gz', col_names = T, delim = '\t'))
top.snps <- add.gtex.annotation(top.snps, gtex, tissue_type = 'breast_gtex_eGenes')

gtex <- suppressMessages(vroom::vroom('~/GERMLINE/GTEx_Analysis_v7_eQTL/Whole_Blood_var_genes.txt.gz', col_names = T, delim = '\t'))
top.snps <- add.gtex.annotation(top.snps, gtex, tissue_type = 'whole_blood_gtex_eGenes')
```

Add SNP consequence 

```{r}
top.snps <- add.consequence(top.snps)
```

Add GWAS catalog annotation

```{r}
gwas <- suppressMessages(vroom::vroom('../../GWAS_catalog/gwas_catalog_v1.0-associations_e98_r2020-02-08.tsv',col_names = T))
top.snps <- add.gwas_catalog(top.snps, gwas)
```


#### Annotated Table

```{r}
top.snps <- top.snps %>% replace(. == 'NA' | is.na(.), '')
top.snps.kable <- top.snps[ , c('snp', 'nearby_gene','nearby_driver','breast_gtex_eGenes','whole_blood_gtex_eGenes','consequence','zscore','susie_pip', 'merged','gwas_catalog' ,'locus')]
colnames(top.snps.kable) <- c('SNP','Nearby Genes (10kb)','Nearby Drivers (100kb)','Breast eGenes','Whole-Blood eGenes','Consequence','ZScore','PIP','Annotation','GWAS Catalog','Locus')
top.snps.kable <- top.snps.kable[order(abs(top.snps.kable$PIP), decreasing = T), ]
kable(top.snps.kable) %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) %>% 
  scroll_box(width = "100%", height = "600px")
```


```{r}
write.table(top.snps.kable, file = '../BRCA/BRCA_finemapped_L1_pip0.1_annotated.txt', sep='\t', col.names = T, row.names = F, quote = F)
```


