---
title: "Gene set enrichment"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align = 'center')
library(ggplot2)
library(gridExtra)
theme_set(theme_bw())

plot.ldsc <- function(h2.partition){

  h2.partition$err <- 1.96*(h2.partition$Enrichment_std_error)
  
  ggplot(h2.partition, aes(x=Category, y=Enrichment, color=Type)) + 
    geom_point() + 
    coord_flip() + 
    geom_hline(yintercept = 1, color="red", linetype="dashed") +
    geom_errorbar(aes(ymin=Enrichment-err, ymax=Enrichment+err), width=.05) +
    theme(text = element_text(size=14),
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank()) +
    ylab('prop h2 / prop of SNPs')
}
```

## GTEx gene set enrichment

### Breast cancer

I obtained ldscores for 54 gene sets based on GTEx expression data from the [Alkas](https://data.broadinstitute.org/alkesgroup/LDSCORE/) group. Below are the ldscore regression results for breast cancer, melanoma, and prostate. The focal tissue is colored in red, while the rest are blue.

Each tissue/gene-set was ran individually, with a control gene set that includes all genes, and the "clean" baseline model with MAF and LD related annotations.

```{r}
brca.gtex <- read.delim('../BRCA/BRCA_GTEx_enrichment_ldsc.txt',header=T, stringsAsFactors = F)
brca.gtex$Type <- ifelse(brca.gtex$Category == 'Breast_Mammary_Tissue', 'Focal Tissue','Nonfocal')
p1 <- plot.ldsc(brca.gtex)
p1 + theme(legend.position='none',axis.text.y = element_text(size=8))
```

### Melanoma

```{r}
mel.gtex <- read.delim('../melanoma//melanoma_GTEx_enrichment_ldsc.txt',header=T, stringsAsFactors = F)
mel.gtex$Type <- ifelse(mel.gtex$Category == 'Skin_Not_Sun_Exposed_(Suprapubic)' | mel.gtex$Category == 'Skin_Sun_Exposed_(Lower_leg)', 'Focal Tissue','Nonfocal')
p1 <- plot.ldsc(mel.gtex)
p1 + theme(legend.position='none',axis.text.y = element_text(size=8))
```

### Prostate cancer

```{r}
prad.gtex <- read.delim('../PRAD/PRAD_GTEx_enrichment_ldsc.txt',header=T, stringsAsFactors = F)
prad.gtex$Type <- ifelse(prad.gtex$Category == 'Prostate', 'Focal Tissue','Nonfocal')
p1 <- plot.ldsc(prad.gtex)
p1 + theme(legend.position='none', axis.text.y = element_text(size=8))
```

## Immune gene sets

I downloaded ldscores for immune gene sets for 290 cell types from the Alkes group. The original source of data is from ImmGen (immunological genome project). There are 290 cell types, but broadly there are 37 categories. Below I have only choosen B cell, dendritic cell, macrophage, natural killer/Tcell, pre-B cell and pre-T cell categories. Within each category are cell types under a range of conditions and growth stages.

### Breast cancer

```{r}
brca.immgen <- read.delim('../BRCA/BRCA_ImmGen_enrichment_ldsc.txt',header=T, stringsAsFactors = F)
type <-sapply(strsplit(brca.immgen$Category, split='[.]'), function(x){return(x[1])})
brca.immgen$Type <- type

brca.immgen <- brca.immgen[brca.immgen$Type %in% c('B','DC','MC','NK','NKT','preB','preT','T'),]

p1 <- plot.ldsc(brca.immgen)
p1 + theme(axis.title.y=element_blank(), axis.text.y=element_blank(), axis.ticks.y=element_blank())
```

### Melanoma

```{r}
mel.immgen <- read.delim('../melanoma/melanoma_ImmGen_enrichment_ldsc.txt',header=T)
mel.immgen$Type <- type
mel.immgen <- mel.immgen[mel.immgen$Type %in% c('B','DC','MC','NK','NKT','preB','preT','T'),]

p1 <- plot.ldsc(mel.immgen)
p1 + theme(axis.title.y=element_blank(),axis.text.y=element_blank(), axis.ticks.y=element_blank())

```

### Prostate

```{r}
prad.immgen <- read.delim('../PRAD/PRAD_ImmGen_enrichment_ldsc.txt',header=T)
prad.immgen$Type <- type
prad.immgen <- prad.immgen[prad.immgen$Type %in% c('B','DC','MC','NK','NKT','preB','preT','T'),]

p1 <- plot.ldsc(prad.immgen)
p1 + theme(axis.title.y=element_blank(), axis.text.y=element_blank(), axis.ticks.y=element_blank())
```
